lib/modules/ens/embarkjs.js

```diff
@@ -229,6 +236,10 @@ __embarkENS.lookup = function (address, callback) {
 __embarkENS.registerSubDomain = function (name, address, callback) {
   callback = callback || function () {};
 
+  if (!web3.eth.defaultAccount) {
+    return callback(defaultAccountNotSetError);
```

Comment:
Shouldn't those be at least instance of `Error`? e.g.

```
return callback(new Error(defaultAccountNotSetError))
```

Reply:
This is a huge pet peeve of mine too, @PascalPrecht. Each module should define their own errors as constants and pass those around :-(

Reply:
Yes, and I think it's enough to first just upgrade this code to use `Error` type in the first place. We can then introduce custom errors in the future, in case we need them.

It's just that, the way it looks like to me right now, is that simply because `callback()` gets called with the message as first parameters, call sides will probably simply expect the first param to be an error a la:

```
doSomething((err, data) => {
  if (err) {
  // whoops
  } 
  // do something with data
}) 
```

But even then `err` is just a string. Semantically it should at least be `Error` IMO. 

For consistency reasons though, I can see why @alaibe went for just using a string. But maybe this is a good opportunity to change to `Error` in all places (in this module).

Reply:
Oh, totally. He's keeping consistent, I'm talking about wishful thinking in the future. Something along the lines of:

```javascript
doSomething((err, data) => {
  if(err) {
    switch(err) {
      case ThisModule.ThisError:
        // ...
    }
  }
});
```

Reply:
It's a bit of a band-aid, but we do have:

https://github.com/embark-framework/embark/blob/develop/lib/utils/utils.js#L484

If an error passed to a callback is destined for logging, it can be run through that utility function if the type isn't known.

Reply:
@michaelsbradleyjr ah I see. But that would in the "best case" still return just a string no? Assuming that case is the `e.message` case. Otherwise `errorMessage()` would also propagate the `Error` instance (which is great!), but then I wonder whether we want a mechanism where we prevent and `Error` to be propagated all the way up.

I think what I'm trying to say is that `err` in the code above would ideally always be some `Error` instance and not a string.
```
doSomething((err, foo) => {
  if (err && err instanceOf Error) { // or instanceOf SomeCustomError
  ...
  }
})
```
And this for all of embark's callback APIs (same would apply to promise-based APIs as well for rejections).

Reply:
@PascalPrecht makes sense, and I do think there is a lot of room for improvement w.r.t. consistency in our implementation. Personally, I'd like to see us move, over time, to a consistent `async` / `await` approach, wrapping any purely callback-based APIs (in 3rd-party packages we depend upon) into promises. The advantage is that try/catch logic in async functions works just like one expects in normal functions, without having to worry about error propagation through callback chains or promise chains. To round it out, [Asynchronous Iterators](https://github.com/tc39/proposal-async-iteration) (already part of the ES 2018 spec, and usable via babel; introduces async generator functions as the basis) can be used to iterate over a series of promises where one step in the series needs to wait on the next.

Reply:
I'm all for that! 👌

Reply:
For completeness-sake this (very old) article pretty much sums up the best practices: https://humanwhocodes.com/blog/2009/03/10/the-art-of-throwing-javascript-errors-part-2/

This can be applied to promise-based/async/await APIs as well. Really what it boils down to is that errors are actual errors, so they can be handled correctly if needed.

Reply:
I am all up for doing changes you suggest here. 
However this is a PR for 3.2 being released very soon.

Also this is a bug fix. I think all these awesome ideas:
- async generators
- await/async
- Error
Deserve at least a real story.
Being that close to 3.2 I just wanted to do the minimum vital change. Not being lazy but just wanted to make sure to break the least possible things.

What do you think?

Reply:
I agree re: min vital change — I consider/ed the other ideas to be for the future, but thought I'd throw in my 2¢.

Reply:
Absolutely! This can be done in the near future.

I've created https://github.com/embark-framework/embark/issues/911 so we don't forget :)